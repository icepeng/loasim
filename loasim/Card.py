from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
