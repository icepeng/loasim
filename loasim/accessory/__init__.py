from loasim.accessory.base import AccessoryStatus
from loasim.accessory.builtin import lostark_accessory_repository
