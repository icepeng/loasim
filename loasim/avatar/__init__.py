from loasim.avatar.base import AvatarState
from loasim.avatar.builtin import lostark_avatar_repository
