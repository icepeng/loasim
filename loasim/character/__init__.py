from loasim.character.base import Character
