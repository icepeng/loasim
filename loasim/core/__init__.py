from loasim.core.buff import (
    BuffManager,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import Skill, SkillSpecification, Tripod
from loasim.core.stat import InternalStat, Stat
