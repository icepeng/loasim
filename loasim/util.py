from typing import Any, Dict

import yaml


def save_config(filename: str, obj: Dict[str, Any]):
    with open(
        filename,
        mode="w",
        encoding="utf8",
    ) as f:
        yaml.safe_dump(obj, f, allow_unicode=True, sort_keys=False)


def load_config(filename: str):
    with open(
        filename,
        encoding="utf8",
    ) as f:
        return yaml.safe_load(f)
