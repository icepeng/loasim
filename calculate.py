from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
