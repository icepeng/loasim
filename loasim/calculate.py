from collections import defaultdict
from typing import Dict, List

from loguru import logger
from pydantic import BaseModel

from loasim.accessory import lostark_accessory_repository
from loasim.avatar import lostark_avatar_repository
from loasim.card import lostark_default_card_repository
from loasim.character.base import Character
from loasim.core import (
    AttackPosition,
    BuffManager,
    BuffState,
    Enemy,
    InternalStat,
    Stat,
)
from loasim.engraving import lostark_engraving_repository
from loasim.gear import lostark_gear_repository
from loasim.job import get_job
from loasim.setitem import lostark_setitem_repository


class SkillAction(BaseModel):
    name: str
    status: Dict[str, BuffState]
    position: AttackPosition


class DealCycle(BaseModel):
    skill_actions: List[SkillAction]
    cycle_time: float


combat_level_map = {
    50: 409,
    51: 410,
    52: 411,
    53: 412,
    54: 414,
    55: 416,
    56: 418,
    57: 420,
    58: 423,
    59: 426,
    60: 429,
}


def get_internal_stat(character: Character) -> InternalStat:
    return (
        InternalStat(
            stat_main=combat_level_map[character.combat_level]
            + character.expedition_stat,
            crit=character.crit,
            special=character.special,
            swift=character.swift,
        )
        + character.bracelet_internal_stat
        + lostark_gear_repository.get_internal_stat(character.gear_status)
        + lostark_accessory_repository.get_internal_stat(character.accessory_status)
        + lostark_avatar_repository.get_internal_stat(character.avatar_status)
    )


def get_basis_stat(
    character: Character, internal_stat: InternalStat, additional_stat: Stat
) -> Stat:
    return (
        internal_stat.get_stat()
        + character.bracelet_stat
        + Stat(pdamage=character.weapon_pdamage)
        + additional_stat
        + lostark_setitem_repository.get_stat(character.setitem_status)
        + lostark_default_card_repository.get_stat(character.card_status)
    )


def calculate(
    character: Character,
    enemy: Enemy,
    deal_cycle: DealCycle,
    additional_stat: Stat = Stat(),
):
    internal_stat = get_internal_stat(character)
    basis_stat = get_basis_stat(character, internal_stat, additional_stat)

    logger.info(internal_stat)
    logger.info(basis_stat)

    job = get_job(character.job_name)
    skills, buffs = job.build(character.skill_status, internal_stat)
    engraving_buffs = lostark_engraving_repository.get_buffs(character.engraving_status)
    setitem_buffs = lostark_setitem_repository.get_buffs(character.setitem_status)
    buff_manager = BuffManager(engraving_buffs + setitem_buffs + buffs)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill_action in deal_cycle.skill_actions:
        skill = skills[skill_action.name]
        buff_stat = buff_manager.get_stat(
            skill_action.status, skill, basis_stat, internal_stat
        )
        damage = skill.get_damage(enemy, skill_action.position, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")

    return total_damage / deal_cycle.cycle_time
