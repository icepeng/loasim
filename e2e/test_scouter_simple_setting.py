from os import path

from loasim.calculate import DealCycle, calculate
from loasim.character import Character
from loasim.core import Enemy
from loasim.util import load_config


def calculate_scouter():
    character = Character.parse_obj(
        load_config(
            path.join(path.dirname(__file__), "configs", "scouter_character.yml")
        )
    )
    deal_cycle = DealCycle.parse_obj(
        load_config(
            path.join(path.dirname(__file__), "configs", "scouter_dealcycle.yml")
        )
    )

    return calculate(
        character=character,
        enemy=Enemy(armor=6000, reduction=23),
        deal_cycle=deal_cycle,
    )


def test_scouter_simple_setting(snapshot: float):
    assert calculate_scouter() == snapshot


if __name__ == "__main__":
    calculate_scouter()
