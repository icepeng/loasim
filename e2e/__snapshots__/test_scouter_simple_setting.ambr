# name: test_scouter_simple_setting
  5598928.878775918
---
